napari/utils/context/_expressions.py

```diff
@@ -0,0 +1,597 @@
+"""This module provides the  :class:`Expr` and its subclasses.
+
+`Expr` is a subclass of `ast.AST` that provides rich dunder methods that
+facilitate joining and comparing typed expressions.  It only implements a
+subset of ast Expressions (for safety of evaluation), but provides more
+than ast.literal_eval.
+
+Expressions that are supported:
+- Names: 'myvar' (these must be evaluated along with some context)
+- Constants: '1'
+- Comparisons: 'myvar > 1'
+- Boolean Operators: 'myvar and yourvar'
+- Binary Operators: 'myvar + 42'
+- Unary Operators: 'not myvar'
+
+Things that are *NOT* supported:
+- attribute access: 'my.attr'
```

Comment:
Thinking at loud.

All the constant like `image_active`, `selection_count` etc will be either defined in napari or plugins, and we don't want attribute access as we want to make sure there is no code execution right ? 

Though it is highly unlikely as we limit what can be in `Constant` for now, do we want to think about some kind of per-plugin namespacing ?

Reply:
ah, this is a good point.  We definitely want per-plugin (and even internal) namespacing, and I _do_ think that we want to support "serialized" context keys with periods in them:

```yaml
commands:
   - command: myplugin.command_name
     when: 'myplugin.some_val > 2'
```

so we could make a special `Name` that allows periods perhaps and fakes it?  But by evaluation time, i think it should be evaluating a pure string/name (without attribute access)

Reply:
I think we can just write a special object that only allow attribute access without allowing properties.

```
class N:
    def __init__(self, **kwargs):
        for k, v in kwargs.items():
            assert isinstance(v, (type(None), str, bytes, bool, int, float, N))
        self.backing = dict(**kwargs)

    def __setattr__(self, key, value):
        if key != 'backing':
            assert isinstance(key, str)
            assert isinstance(
                value, (type(None), str, bytes, bool, int, float, N)
            )
            self.backing[key] = value
        else:
            return super().__setattr__(key, value)

    def __getattr__(self, key):
        if key != 'backing':
            return self.backing[key]

```

? Seem to work locally. Also with do want `type(x) in {...}` instead of isinstance to avoid subclassing

Reply:
Thoughts about https://github.com/Carreau/napari/pull/new/expressions  / 38a11f4e164b13d167911bd8ef49045cca543141 ?

Reply:
interesting.  I like it! ... I'm inclined to leave it for later though, is that ok?

Reply:
Yes, sure, I'm more exploring the possibilities, I think that flat is good for a first pass.
And doing this to see if I can abuse it.

Reply:
regarding the parens there... I think the only thing I don't love is that it will add parens even to simple expressions ...  so all of the serialization will be a bit uglier by default, right?

Reply:
> And doing this to see if I can abuse it.

yay :) thank you!

Reply:
> regarding the parens there... I think the only thing I don't love is that it will add parens even to simple expressions ... so all of the serialization will be a bit uglier by default, right?

You can easily remove the top level parenthesis:

```python
def visit(self, item, *,  parenthesise=True):
        if isinstance(item, ast.AST):
            if item.__class__.__name__.endswith(
                ('Op', 'Compare', 'Exp')
            ) and parenthesise:
                self.write('(')
                super().visit(item)
                self.write(')')
            else:
                super().visit(item)

    
    def __init__(self,...):
        ...
        if node is not None:
            self.visit(node, parenthesise=False)
```

```python
>>> for x in ['a + 1', 'a.x > 3', 'layer.active and layer.ndim<3', 'a+b == c', 'a and b',
...     'a == 1',
...     'a if b == 7 else False',
...     '1',
...     '3.14',
...     'True',
...     'False',
...     'None',
...     'hieee',
...     "b'bytes'"]:
...     print(normalise(x))
a + 1
a.x > 3
layer.active and (layer.ndim < 3)  # ok a bit worse
(a + b) == c  # ok a bit worse as well
a and b
a == 1
a if (b == 7) else False # and not perfect.
1
3.14
True
False
None
hieee
b'bytes'.  #< we def want to check strings as well.
```

Reply:
> You can easily remove the top level parenthesis:

oh right of course!